src/audio/crossover/crossover.c

```diff
@@ -642,215 +562,180 @@ static int crossover_copy(struct comp_dev *dev)
-
-out:
-	buffer_release(source_c);
-
 	return ret;
 }
 
+#if CONFIG_IPC_MAJOR_4
+static void crossover_set_stream_params(struct comp_buffer *buffer, struct processing_module *mod)
+{
+	struct comp_buffer __sparse_cache *buffer_c;
+	enum sof_ipc_frame valid_fmt, frame_fmt;
+
+	buffer_c = buffer_acquire(buffer);
+	audio_stream_fmt_conversion(mod->priv.cfg.base_cfg.audio_fmt.depth,
+				    mod->priv.cfg.base_cfg.audio_fmt.valid_bit_depth,
+				    &frame_fmt, &valid_fmt,
+				    mod->priv.cfg.base_cfg.audio_fmt.s_type);
+	audio_stream_set_buffer_fmt(&buffer_c->stream,
+				    mod->priv.cfg.base_cfg.audio_fmt.interleaving_style);
+	audio_stream_set_channels(&buffer_c->stream,
+				  mod->priv.cfg.base_cfg.audio_fmt.channels_count);
+	audio_stream_set_rate(&buffer_c->stream,
+			      mod->priv.cfg.base_cfg.audio_fmt.sampling_frequency);
+	audio_stream_set_frm_fmt(&buffer_c->stream, frame_fmt);
+	audio_stream_set_valid_fmt(&buffer_c->stream, valid_fmt);
+	buffer_release(buffer_c);
+}
+#endif
+
 /**
  * \brief Prepares Crossover Filter component for processing.
  * \param[in,out] dev Crossover Filter base component device.
  * \return Error code.
  */
-static int crossover_prepare(struct comp_dev *dev)
+static int crossover_prepare(struct processing_module *mod,
+			     struct sof_source __sparse_cache **sources, int num_of_sources,
+			     struct sof_sink __sparse_cache **sinks, int num_of_sinks)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_data *cd =  module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
 	struct comp_buffer *source, *sink;
 	struct comp_buffer __sparse_cache *source_c, *sink_c;
 	struct list_item *sink_list;
-	int32_t sink_period_bytes;
-	int ret;
+	int channels;
+	int ret = 0;
 
 	comp_info(dev, "crossover_prepare()");
 
-	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
-	if (ret < 0)
-		return ret;
-
-	if (ret == COMP_STATUS_STATE_ALREADY_SET)
-		return PPL_STATUS_PATH_STOP;
-
 	/* Crossover has a variable number of sinks */
-	source = list_first_item(&dev->bsource_list,
-				 struct comp_buffer, sink_list);
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
 	source_c = buffer_acquire(source);
 
 	/* Get source data format */
 	cd->source_format = audio_stream_get_frm_fmt(&source_c->stream);
+	channels = audio_stream_get_channels(&source_c->stream);
+	audio_stream_init_alignment_constants(1, 1, &source_c->stream);
+	buffer_release(source_c);
+
+#if CONFIG_IPC_MAJOR_4
+	list_for_item(sink_list, &dev->bsink_list) {
+		sink = container_of(sink_list, struct comp_buffer, source_list);
+		crossover_set_stream_params(sink, mod);
+	}
+#endif
 
 	/* Validate frame format and buffer size of sinks */
 	list_for_item(sink_list, &dev->bsink_list) {
 		sink = container_of(sink_list, struct comp_buffer, source_list);
 		sink_c = buffer_acquire(sink);
 
+		/* Note: Printing pipeline IDs for debug, these are with IPC4 zeros, probably due
```

Comment:
Does the bug you mentioned still exist under IPC4? If yes, is it worked-around by the commit, or is it non-critical?

Reply:
I'm now after vacation starting to study this again -- I hope I'll gain understanding of IPC4 to be able to address this.